File: leetcode-go/Q001-Q100/Q037-Sudoku-Solver/Q037-Sudoku-Solver.go
package Q037_Sudoku_Solver

var (
	chars  = []byte{'1', '2', '3', '4', '5', '6', '7', '8', '9'}
)

func solveSudoku(board [][]byte) {
	dfs(board, 0)
}

func dfs(board [][]byte, pos int) bool {
	if pos >= 81 {
		return true
	}

	i, j := pos/9, pos%9
	if board[i][j] != '.' {
		return dfs(board, pos+1)
	} else {
		for _, c := range chars {
			if isValid(board, i, j, c) {
				board[i][j] = c
				if dfs(board, pos+1) {
					return true
				} else {
					board[i][j] = '.'
				}
			}
		}
	}
	return false
}

func isValid(board [][]byte, i, j int, c byte) bool {
	for k := 0; k < 9 ; k++ {
		if board[i][k] == c {
			return false
		}
		if board[k][j] == c {
			return false
		}

		row, col := i/3, j/3
		for m := 3*row; m < 3*(row+1); m++ {
			for n := 3*col; n < 3*(col+1); n++ {
				if board[m][n] == c {
					return false
				}
			}
		}

	}
	return true
}

File: src/Y2018/backtracking/Q037_Sudoku_Solver.java
package Y2018.backtracking;

public class Q037_Sudoku_Solver {

    public void solveSudoku(char[][] board) {
        if (board == null || board.length == 0 || board[0] == null || board[0].length == 0) {
            return;
        }
        solve(board);
    }

    private boolean solve(char[][] board) {
        int m = board.length, n = board[0].length;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] != '.') {
                    continue;
                }
                for (char c = '1'; c <= '9'; c++) {
                    if (isValid(board, i, j, c)) {
                        board[i][j] = c;
                        if (solve(board)) {
                            return true;
                        }
                        board[i][j] = '.';
                    }
                }
                return false;
            }
        }
        return true;
    }

    private boolean isValid(char[][] board, int row, int col, char c) {
        for (int i = 0; i < 9; i++) {
            if (board[i][col] != '.' && board[i][col] == c) return false;
            if (board[row][i] != '.' && board[row][i] == c) return false;
            if (board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] != '.' &&
                    board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == c) return false;
        }
        return true;
    }

}
